test/conformance/runtime/liveness_probe_test.go

```diff
@@ -0,0 +1,180 @@
+
+	if err != nil {
+		t.Fatalf("Failed to create initial Service: %v: %v", names.Service, err)
+	}
+
+	// Wait for the liveness probe to be executed a few times before introducing a failure.
+	url := resources.Route.Status.URL.URL()
+	url.Path = livenessCounterPath
+	if _, err = pkgtest.WaitForEndpointState(
+		context.Background(),
+		clients.KubeClient,
+		t.Logf,
+		url,
+		AtLeastNumLivenessChecks(t, 5),
+		"livenessIsReady",
+		test.ServingFlags.ResolvableDomain,
+		test.AddRootCAtoTransport(context.Background(), t.Logf, clients, test.ServingFlags.HTTPS),
+	); err != nil {
+		t.Fatalf("The endpoint for Route %s at %s didn't return success: %v", names.Route, url, err)
+	}
+
+	// Check that user-container hasn't been restarted yet.
+	deploymentName := resourcenames.Deployment(resources.Revision)
+	podList, err := clients.KubeClient.CoreV1().Pods(test.ServingFlags.TestNamespace).List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		t.Fatal("Unable to get pod list: ", err)
+	}
+	for i := range podList.Items {
+		pod := &podList.Items[i]
+		if strings.Contains(pod.Name, deploymentName) && UserContainerRestarted(pod) {
+			t.Fatal("User container unexpectedly restarted")
+		}
+	}
+
+	t.Log("POST to /start-failing")
+	client, err := pkgtest.NewSpoofingClient(context.Background(),
+		clients.KubeClient,
+		t.Logf,
+		url.Hostname(),
+		test.ServingFlags.ResolvableDomain,
+		test.AddRootCAtoTransport(context.Background(), t.Logf, clients, test.ServingFlags.HTTPS))
+	if err != nil {
+		t.Fatalf("Failed to create spoofing client: %v", err)
+	}
+
+	url.Path = "/start-failing"
+	startFailing, err := http.NewRequest(http.MethodPost, url.String(), nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, err := client.Do(startFailing); err != nil {
+		t.Fatalf("POST to /start-failing failed: %v", err)
+	}
+
+	// Wait for the user-container to be restarted.
+	if err := pkgtest.WaitForPodListState(
+		context.Background(),
+		clients.KubeClient,
+		func(p *corev1.PodList) (bool, error) {
+			for i := range p.Items {
+				pod := &p.Items[i]
+				if strings.Contains(pod.Name, deploymentName) && UserContainerRestarted(pod) {
+					return true, nil
+				}
+			}
+			return false, nil
+		},
+		"WaitForContainerRestart", test.ServingFlags.TestNamespace); err != nil {
+		t.Fatalf("Failed waiting for user-container to be restarted: %v", err)
+	}
+
+	// After restart, verify the liveness probe passes a few times again.
+	url.Path = livenessCounterPath
+	if _, err = pkgtest.WaitForEndpointState(
+		context.Background(),
+		clients.KubeClient,
+		t.Logf,
+		url,
+		AtLeastNumLivenessChecks(t, 5),
+		"livenessIsReady",
+		test.ServingFlags.ResolvableDomain,
+		test.AddRootCAtoTransport(context.Background(), t.Logf, clients, test.ServingFlags.HTTPS),
+	); err != nil {
+		t.Fatalf("The endpoint for Route %s at %s didn't return success: %v", names.Route, url, err)
+	}
+}
+
+func AtLeastNumLivenessChecks(t *testing.T, expectedChecks int) spoof.ResponseChecker {
```

Comment:
Can these be lowercase? Or moved to e2e.go if shared?

Reply:
I can make them lowercase. Same for UserContainerRestarted below.